compiler/rustc_codegen_ssa/src/mir/naked_asm.rs

```diff
@@ -0,0 +1,242 @@
+        };
+
+        let operands: Vec<_> =
+            operands.iter().map(|op| self.inline_to_global_operand(op)).collect();
+
+        let item_data = cx.codegen_unit().items().get(&MonoItem::Fn(instance)).unwrap();
+        let (begin, end) = crate::mir::naked_asm::prefix_and_suffix(cx.tcx(), instance, item_data);
+
+        let mut template_vec = Vec::new();
+        template_vec.push(rustc_ast::ast::InlineAsmTemplatePiece::String(begin));
+        template_vec.extend(template.iter().cloned());
+        template_vec.push(rustc_ast::ast::InlineAsmTemplatePiece::String(end));
+
+        cx.codegen_global_asm(&template_vec, &operands, options, line_spans);
+    }
+
+    fn inline_to_global_operand(&self, op: &InlineAsmOperand<'tcx>) -> GlobalAsmOperandRef<'tcx> {
+        match op {
+            InlineAsmOperand::Const { value } => {
+                let const_value = self.eval_mir_constant(value);
+                let string = common::asm_const_to_str(
+                    self.cx.tcx(),
+                    value.span,
+                    const_value,
+                    self.cx.layout_of(value.ty()),
+                );
+                GlobalAsmOperandRef::Const { string }
+            }
+            InlineAsmOperand::SymFn { value } => {
+                let instance = match value.ty().kind() {
+                    &ty::FnDef(def_id, args) => Instance::new(def_id, args),
+                    _ => bug!("asm sym is not a function"),
+                };
+
+                GlobalAsmOperandRef::SymFn { instance }
+            }
+            InlineAsmOperand::SymStatic { def_id } => {
+                GlobalAsmOperandRef::SymStatic { def_id: *def_id }
+            }
+            InlineAsmOperand::In { .. }
+            | InlineAsmOperand::Out { .. }
+            | InlineAsmOperand::InOut { .. }
+            | InlineAsmOperand::Label { .. } => {
+                bug!("invalid operand type for naked_asm!")
+            }
+        }
+    }
+}
+
+enum AsmBinaryFormat {
+    Elf,
+    Macho,
+    Coff,
+}
+
+impl AsmBinaryFormat {
+    fn from_target(target: &rustc_target::spec::Target) -> Self {
+        if target.is_like_windows {
+            Self::Coff
+        } else if target.options.vendor == "apple" {
+            Self::Macho
+        } else {
+            Self::Elf
+        }
+    }
+}
+
+fn linkage_directive(linkage: Linkage) -> Option<&'static str> {
+    if true {
+        // this is required. emitting nothing or .weak will emit
+        //
+        // > Global is external, but doesn't have external or weak linkage!
+        //
+        // and then aborts compilation
+        return Some(".globl");
+    }
+
+    match linkage {
+        Linkage::External => Some(".globl"),
+        Linkage::WeakAny | Linkage::WeakODR => Some(".weak"),
+        Linkage::LinkOnceAny | Linkage::LinkOnceODR => Some(".weak"),
+        Linkage::Internal | Linkage::Private => None, // just doesn't emit any attribute
```

Comment:
cg_clif emits assembly in a separate object file, so everything define in assembly that can be called from rust needs to be exported from the assembly object file.

Reply:
By the way this is what cg_clif uses for translating naked functions to `global_asm!`: https://github.com/rust-lang/rustc_codegen_cranelift/blob/b929b68a5ba4570e15637b12d9e687a2ba6673d8/src/inline_asm.rs#L208-L315

Reply:
yes, that code was very helpful. What is generated here is based on https://github.com/Amanieu/naked-function which is more complicated, but also supports more architectures. it would be nice of course if these could be merged.